# **Feature engineering - liczebność cech**

In [9]:
import pandas as pd
import numpy as np

In [10]:
data = pd.read_csv('C:/Users/48692/Desktop/ML/Lab3/data.csv')

### 1. Sprawdź liczebność poszczególnych etykiet dla danych zmiennych jakościowych

In [11]:
var_names = list(data.columns[0:2]) + list(data.columns[3:])
for var_name in var_names:
  print('Liczba etykiet zmiennej {}: {}'.format(var_name, len(data[var_name].unique())))

Liczba etykiet zmiennej pclass: 3
Liczba etykiet zmiennej survived: 2
Liczba etykiet zmiennej sex: 2
Liczba etykiet zmiennej age: 99
Liczba etykiet zmiennej sibsp: 7
Liczba etykiet zmiennej parch: 8
Liczba etykiet zmiennej ticket: 929
Liczba etykiet zmiennej fare: 282
Liczba etykiet zmiennej cabin: 187
Liczba etykiet zmiennej embarked: 4
Liczba etykiet zmiennej boat: 28
Liczba etykiet zmiennej body: 122
Liczba etykiet zmiennej home.dest: 370


### 2. Wyświetl z użyciem funkcji print liczbę wszystkich pasażerów. Wykorzystaj podobny sposób jak w punkcie 1. 

In [12]:
print('Liczba wszystkich pasażerów: {}'.format(len(data)))

Liczba wszystkich pasażerów: 1309


### 3. Skomentuj wyniki otrzymane w punkcie 1 i 2. Podziel zmienne ze względu na dużą i małą moc zbioru (kardynalność). 

Wyniki otrzymane dla poszczególnych zmiennych są zróżnicowane ze względu na kardynalność - część zmiennych ma bardzo dużo etykiet, a inne bardzo mało. Najwięcej etykiet zaobserwowano dla zmiennej ticket, bo aż 929, przy czym liczba wszystkich pasażerów wyniosła 1309.

**Podział zmiennych ze względu na moc zbioru:**
* mała moc zbioru: *pclass, survived, sex, sibsp, parch, embarked, boat*
* duża moc zbioru: *age, ticket, fare, cabin, body, home.dest*


### 4. Sprawdź, ile unikalnych etykiet ma zmienna mówiąca o kabinie danego pasażera. Użyj takiej funkcji, która zwraca wynik w postaci NumPy array.

In [13]:
print('Liczba unikalnych etykiet dla zmiennej cabin: {}'.format(len((data.cabin.unique()))))
print('Tablica unikalnych etykiet dla zmiennej cabin:\n {}'.format(np.unique(data.cabin)))

Liczba unikalnych etykiet dla zmiennej cabin: 187
Tablica unikalnych etykiet dla zmiennej cabin:
 ['?' 'A10' 'A11' 'A14' 'A16' 'A18' 'A19' 'A20' 'A21' 'A23' 'A24' 'A26'
 'A29' 'A31' 'A32' 'A34' 'A36' 'A5' 'A6' 'A7' 'A9' 'B10' 'B101' 'B102'
 'B11' 'B18' 'B19' 'B20' 'B22' 'B24' 'B26' 'B28' 'B3' 'B30' 'B35' 'B36'
 'B37' 'B38' 'B39' 'B4' 'B41' 'B42' 'B45' 'B49' 'B5' 'B50' 'B51 B53 B55'
 'B52 B54 B56' 'B57 B59 B63 B66' 'B58 B60' 'B61' 'B69' 'B71' 'B73' 'B77'
 'B78' 'B79' 'B80' 'B82 B84' 'B86' 'B94' 'B96 B98' 'C101' 'C103' 'C104'
 'C105' 'C106' 'C110' 'C111' 'C116' 'C118' 'C123' 'C124' 'C125' 'C126'
 'C128' 'C130' 'C132' 'C148' 'C2' 'C22 C26' 'C23 C25 C27' 'C28' 'C30'
 'C31' 'C32' 'C39' 'C45' 'C46' 'C47' 'C49' 'C50' 'C51' 'C52' 'C53' 'C54'
 'C55 C57' 'C6' 'C62 C64' 'C65' 'C68' 'C7' 'C70' 'C78' 'C80' 'C82' 'C83'
 'C85' 'C86' 'C87' 'C89' 'C90' 'C91' 'C92' 'C93' 'C95' 'C97' 'C99' 'D'
 'D10 D12' 'D11' 'D15' 'D17' 'D19' 'D20' 'D21' 'D22' 'D26' 'D28' 'D30'
 'D33' 'D34' 'D35' 'D36' 'D37' 'D38' 'D40

### 5. Zredukuj liczbę cech dla zmiennej opisującej kabiny poprzez zastąpienie obecnych etykiet w formacie PP11 do etykiet zawierających tylko pierwszą literę. Użyj astype(str).str[pozycja]. Nową zmienną nazwij CabinReduced. Wyświetl pierwsze 25 wierszy zbioru danych dla kolumn Cabin i CabinReduced.


In [14]:
data['CabinReduced'] = data.cabin.astype(str).str[0]
print(data[['cabin', 'CabinReduced']].head(25))

      cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9         ?            ?
10  C62 C64            C
11  C62 C64            C
12      B35            B
13        ?            ?
14      A23            A
15        ?            ?
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C
20      D35            D
21      D35            D
22     C148            C
23        ?            ?
24      C97            C


### 6. Wyświetl (jak w pkt 1) liczbę etykiet dla zmiennych z punkt 5. O ile procent zredukowano kardynalność zbioru zmiennej opisującej kabiny?

In [15]:
print('Liczba etykiet zmiennej cabin: {}\nLiczba etykiet zmiennej CabinReduced: {}'.format(len(data.cabin.unique()), len(data.CabinReduced.unique())))

Liczba etykiet zmiennej cabin: 187
Liczba etykiet zmiennej CabinReduced: 9


In [16]:
print('Kardynalność zbioru zmiennej opisującej kabiny zredukowano o: {}%'.format(round(100-(9/187*100), 2)))

Kardynalność zbioru zmiennej opisującej kabiny zredukowano o: 95.19%


### 7. Uzasadnij, dlaczego dokonujesz redukcji akurat tej zmiennej. Jak to wpływa na przyszłe analizy. Czy powoduje jakieś negatywne skutki?

Redukcja została dokonana dla zmiennej *cabin* ze względu na dużą moc zbioru (187 etykiet). Generalizacja tej zmiennej jest wskazana, gdyż można zredukować etykiety w ten sposób, aby określały jedynie przynależność do konkretnej strefy/typu kabiny, dzięki czemu przyszły model będzie bardziej uniwersalny, a wciąż będzie dobrze sobie radził. Do negatywnych skutków można zaliczyć utratę informacji na temat konktretnego pokoju przydzielonego dla danego pasażera, zatem nie będzie możliwości stworzenia modelu, który prawidłowo przewidywałby, w którym pokoju podróżowała wybrana osoba.